docs/question/server.md

```diff
@@ -1,89 +1,312 @@
+```
+
+!!! info
+
+    In general, each function of the question generation process runs *after* all elements in the question. For example, the `parse()` function in `server.py` runs after all elements have run their `parse()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.
+
+Although questions with custom grading usually don't use the grading functions from individual elements, it is _highly_ recommended that built-in elements are used for student input, as these elements include helpful parsing and feedback by default. Parsed student answers are present in the `data["submitted_answers"]` dictionary.
+
+!!! note
+
+    Data stored under the `"submitted_answers"` key in the data dictionary may be of varying type. Specifically, the `pl-integer-input`
+    element sometimes stores very large integers as strings instead of the Python `int` type used in most cases. The best practice for custom grader
+    code in this case is to always cast the data to the desired type, for example `int(data["submitted_answers"][name])`. See the
+    [PrairieLearn elements documentation](../elements.md) for more detailed discussion related to specific elements.
+
+## Step 5: `grade`
+
+Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:
+
+- Setting the score and feedback for each named answer in `data["partial_scores"]`.
+- Setting the total score for the question in `data["score"]`.
+- Setting the overall feedback for the question in `data["feedback"]`.
+
+This function only runs if `parse()` did not produce format errors, so we can assume all data is valid. All elements will have already graded their answers (if any) before this point.
+
+- `data["partial_scores"][NAME]["score"]` is the individual scores for the named answers (0 to 1).
+- `data["partial_scores"][NAME]["feedback"]` is the feedback for each named answer from the element that graded it.
+- `data["score"]` is the total score for the question (0 to 1).
+- `data["feedback"][NAME]` is the overall question feedback for each named answer.
+
+It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.
+
+If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer. In the snipped below, we update `data["score"]` using the [`set_weighted_score_data`][prairielearn.question_utils.set_weighted_score_data] utility.
```

Comment:
This last statement probably deserves a paragraph of its own, explaining why it's necessary (the grade function has the option of having the combined score of the question be computed from the partial scores or be overridden with a custom total) and that it's a 0-1 score.